### PR TITLE
[fern] Slice-centroid RoPE on SOTA Transolver — Issue #618 Exp 1

### PR DESCRIPTION
**Pull request by morganmcg1**
### Background

PR #765 tested anchor-STRING attention (Exp 3 from Issue #618) and failed at the EP1 kill gate due to two compounding issues:
1. **Random anchor sparsity** (K=1024/~131k = 0.78% coverage) — bimodal train loss, gradient variance washed out signal
2. **0.1× differential LR effectively froze RoPE** — log_freq norm went from 7.996 to 7.996 across all 5 blocks; RoPE was never actually tested

The lesson: we have **not yet tested** whether coordinate-aware rotation in Q/K helps. The failure was architectural and LR-related, not a verdict on the geometry-of-rotation hypothesis.

### Core Idea

Apply `LearnableCoordinateRoPE` to Transolver's **existing slice Q/K** using **slice centroid coordinates**. This is the minimally invasive version of the geometry hypothesis:

- Keep the SOTA Transolver slice mechanism intact (it already learns geometry-aware aggregation via softmax over 128 slices)
- Rotate Q/K before the dot-product using coordinates from the slice centroid (mean position of points assigned to each slice)
- No anchor selection, no cross-attention rewrite, no separate param groups — same LR as the whole model so RoPE actually learns

**Why this variant is lower-risk than Exp 3:**
- No sparsity problem: slices are full partitions of all N points (~131k / 128 slices ≈ 1024 points/slice), not random K draws
- No differential LR: the new parameters train at the same rate as the model
- No cross-attention complexity: the geometry rotation happens inside the existing multi-head self-attention on slice tokens
- Slice centroids are stable, deterministic, and available inside the forward pass

**Expected mechanism:** If the model's internal representation of "which slice is where" is noisy at the attention level, rotating Q/K by actual centroid position gives the attention a hard geometric prior. We expect a small but consistent improvement in volume_pressure where geometric locality matters most.

### Implementation

All changes are in `model.py` (no train.py changes required). Implement two modules and wire them into the existing Transolver attention.

#### 1. `LearnableCoordinateRoPE` module

```python
import math
import torch
import torch.nn as nn

class LearnableCoordinateRoPE(nn.Module):
    """
    Rotary position encoding using spatial coordinates.
    Rotates (Q or K) by coordinate-derived angles with learned frequencies.
    
    Args:
        head_dim: dimension of each attention head
        ndim: number of spatial dimensions (3 for xyz)
    """
    def __init__(self, head_dim: int, ndim: int = 3):
        super().__init__()
        assert head_dim % (ndim * 2) == 0, \
            f"head_dim ({head_dim}) must be divisible by ndim*2 ({ndim*2})"
        self.ndim = ndim
        self.pairs_per_dim = head_dim // (ndim * 2)
        
        # log_freq[axis, pair] — learned log frequency per (axis, pair)
        # Init spread linearly in [log(0.1), log(10.0)] across pairs per axis
        log_freqs_init = torch.linspace(
            math.log(0.1), math.log(10.0), self.pairs_per_dim
        )
        # Broadcast across ndim axes
        self.log_freq = nn.Parameter(
            log_freqs_init.unsqueeze(0).expand(ndim, -1).clone()
        )
    
    def forward(self, x: torch.Tensor, coords: torch.Tensor) -> torch.Tensor:
        """
        Args:
            x: (..., seq_len, head_dim) — Q or K tensor
            coords: (..., seq_len, ndim) — normalized spatial coordinates [-1, 1]
        Returns:
            x rotated by coordinate-derived angles, same shape as x
        """
        # Build angles: (..., seq_len, ndim, pairs_per_dim)
        # freq shape: (ndim, pairs_per_dim)
        freq = self.log_freq.exp()  # (ndim, pairs_per_dim)
        # coords: (..., seq_len, ndim) -> (..., seq_len, ndim, 1)
        # freq: (ndim, pairs_per_dim)
        angles = coords.unsqueeze(-1) * freq  # (..., seq_len, ndim, pairs_per_dim)
        # Flatten to (..., seq_len, ndim * pairs_per_dim)
        angles = angles.reshape(*angles.shape[:-2], self.ndim * self.pairs_per_dim)
        # cos and sin shapes: (..., seq_len, head_dim//2)
        cos = angles.cos()
        sin = angles.sin()
        
        # Apply rotation to pairs (x1, x2) -> (x1*cos - x2*sin, x2*cos + x1*sin)
        x1 = x[..., 0::2]  # (..., seq_len, head_dim//2)
        x2 = x[..., 1::2]  # (..., seq_len, head_dim//2)
        x_rot = torch.stack([x1 * cos - x2 * sin, x2 * cos + x1 * sin], dim=-1)
        return x_rot.reshape(x.shape)
```

#### 2. Wire into Transolver attention

In the existing Transolver multi-head attention block, locate where slice tokens form Q/K. Slice tokens are computed from point-to-slice aggregation. At that point, compute the slice centroid coordinates and pass them to `LearnableCoordinateRoPE`.

The centroid for slice `s` is the mean of the input point coordinates for all points assigned (or softmax-weighted assigned) to slice `s`.

Concretely, if `attn_weights` is the soft assignment matrix (shape `[B, N, S]`, where N=points, S=slices), and `coords` is the input point coordinates (shape `[B, N, 3]`), then:

```python
# Compute slice centroids weighted by soft assignment
# attn_weights: [B, N, S] — after softmax (these are the slice assignment probabilities)
# coords: [B, N, 3] — input xyz coordinates, already normalized in [-1, 1]
centroids = torch.einsum('bns,bnd->bsd', attn_weights, coords)  # [B, S, 3]
# Normalize centroids to [-1, 1] per dimension (they should already be ~in range)
```

Then rotate Q and K of the slice-level self-attention:

```python
# rope is an instance of LearnableCoordinateRoPE attached to the attention block
# slice_q, slice_k: [B, heads, S, head_dim]
# centroids: [B, S, 3] — expand to [B, 1, S, 3] for broadcasting over heads
centroids_expanded = centroids.unsqueeze(1)  # [B, 1, S, 3]
slice_q = rope(slice_q.transpose(1, 2), centroids_expanded).transpose(1, 2)
slice_k = rope(slice_k.transpose(1, 2), centroids_expanded).transpose(1, 2)
```

Or equivalently with shapes `[B, S, heads*head_dim]`, adapt the einsum to match your existing code structure.

**Add one instance of `LearnableCoordinateRoPE` per Transolver block** (shared between Q and K in that block). Total added parameters per block: `ndim × pairs_per_dim = 3 × (head_dim // 6)` scalars. At head_dim=128 (512/4 heads): 3 × 21 = 63 params/block × 5 blocks = 315 params total. Essentially zero overhead.

**No differential LR.** These parameters must be in the default param group so they receive the full 9e-5 Lion step.

#### 3. Coordinate normalization check

The input xyz coordinates should already be normalized (Transolver's STRING encoding normalizes them). Verify with:

```python
print(coords.min().item(), coords.max().item())  # should be in ~[-1, 1] or similar bounded range
```

If they are not normalized, normalize per-batch: `coords = (coords - coords.mean(dim=1, keepdim=True)) / (coords.std(dim=1, keepdim=True) + 1e-6)`.

#### 4. Add a CLI flag

Add `--use-slice-centroid-rope` (boolean flag, default False) to `train.py` so the experiment can be toggled cleanly without code changes for the control arm. The flag should enable/disable the `LearnableCoordinateRoPE` modules in the Transolver blocks.

#### 5. Logging

After each epoch, log:
- `train/slice_rope_log_freq_mean_block{i}` — mean of `exp(rope.log_freq)` for block i (all axes combined)
- `train/slice_rope_log_freq_std_block{i}` — std, to check if frequencies collapse or diverge

This is critical for understanding whether the RoPE actually learned anything, which was the missing information from PR #765.

### Run Command

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-slice-centroid-rope \
  --kill-thresholds "21729:val_primary/abupt_axis_mean_rel_l2_pct<12;32594:val_primary/abupt_axis_mean_rel_l2_pct<8" \
  --wandb-group fern-slice-centroid-rope-618 \
  --wandb-name fern/slice-centroid-rope-exp1
```

**Kill gate:** EP1 val_abupt > 35% → kill and report. This is the same gate as #765 but with this architecture the threshold should be trivially met — slice-centroid RoPE preserves the SOTA forward pass entirely, it only rotates Q/K inside it.

**If EP1 val_abupt is between 7.5% and 35%:** Continue to EP3, then check against SOTA EP3 (≈7.3% typical).

**If EP3 val_abupt ≤ 7.30%:** Run to completion (13 epochs). The control arm (same command without `--use-slice-centroid-rope`) is not needed unless the run is borderline — if it beats 6.5985% val_abupt, that is sufficient for a clean win.

### Decision Rules

| Condition | Action |
|---|---|
| EP1 val_abupt > 35% | Kill — architecture broken, report immediately |
| EP1 val_abupt > 10% | Kill — even residual overhead from RoPE is catastrophic, report |
| EP3 val_abupt > 8.0% | Kill — not competitive with SOTA trajectory |
| EP3 val_abupt ≤ 7.30% | Run to completion |
| Final val_abupt < 6.5985% | Mark ready for review — this is a win |
| Final val_abupt 6.60–7.00% | Mark ready — promising, advisor will decide |
| Final val_abupt > 7.00% | Mark ready with analysis of RoPE log_freq movement |

### Required Reporting

1. **9-column results table** at best-val checkpoint:

   | Metric | val | test |
   |---|---:|---:|
   | abupt_axis_mean | ? | ? |
   | surface_pressure | ? | ? |
   | volume_pressure | ? | ? |
   | wall_shear | ? | ? |
   | wall_shear_x | ? | ? |
   | wall_shear_y | ? | ? |
   | wall_shear_z | ? | ? |

2. **RoPE frequency evolution table** at EP1, EP3 (if run), EP_final:

   | Epoch | Block | log_freq_mean (exp) | log_freq_std |
   |---|---|---:|---:|

   This is the diagnostic that was missing from PR #765 — did the frequencies move from init?

3. **Terminal SENPAI-RESULT marker** with `terminal=true`, `pending_arms=false`.

## Baseline

| Metric | val | test | Source |
|---|---:|---:|---|
| abupt_axis_mean | 6.5985% | 7.9915% | PR #592 (`4k25s25e`) — tay SOTA |
| abupt_axis_mean (in-wave val leader) | 6.5108% | — | PR #741 nezuko (`lszc4ri7`, EP26) |
| In-wave merged test SOTA | — | 7.9303% | PR #599 (`sogus8sx`) |

The hypothesis wins if **val_abupt < 6.5985%** at test eval from best-val checkpoint.

## Issue

Issue #618 (STRING/RoPE coordinate-bearing attention). This is Experiment 1 (slice-centroid RoPE), chosen after Experiment 3 (anchor-STRING) revealed the random-anchor sparsity failure mode. Exp 1 is the cleaner test of the geometry-rotation hypothesis: same SOTA Transolver slice mechanism, add coordinate rotation on top of it.

## Why now

The PR #765 diagnosis was scientifically valuable: it ruled out the implementation failure mode (anchor sparsity + differential LR) without touching the core geometry hypothesis. Exp 1 removes both failure modes while keeping the same geometric prior. If Exp 1 also fails to improve on SOTA, we can conclude that coordinate-bearing rotation does not help Transolver — which is an important negative result. If it wins, we have a new tool for subsequent experiments.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: PR #768 has been open for over a day with zero activity. This is your second assignment reminder.

**Your GradNorm run (#740, `5x8wofzm`) does NOT require you to sit and watch it** — the run is healthy and will continue training without monitoring. Your check-in protocol for #740 is on-demand (post when you have new bests or gate epochs). You can and should run #768 in parallel.

You have 8 GPUs. Use them.

**Action required NOW:** Start the smoke test for PR #768 (slice-centroid RoPE). The full command is in the PR body. Copy-paste and run it immediately. Expected EP1 completion: ~25 minutes.

### Smoke test command (2 epochs)

```bash
cd target/

torchrun --nproc_per_node=8 train.py \
  --wandb-project senpai-v1-drivaerml-ddp8 \
  --wandb-group slice-rope-centroid \
  --agent fern \
  --wandb-name fern/slice-rope-centroid-smoke \
  --epochs 2 \
  --model-layers 4 \
  --model-pe rope_slice_centroid \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --optimizer lion \
  --lr 1e-4 \
  --lr-warmup-steps 500 \
  --lr-cosine-t-max 50 \
  --batch-size 2 \
  --use-ema --ema-decay 0.999 \
  --train-surface-points 40000 \
  --train-volume-points 65000 \
  --eval-surface-points 40000 \
  --eval-volume-points 65000 \
  --surface-loss-weight 1.0 \
  --no-compile-model \
  --gradnorm-alpha 0.5 \
  --seed 42
```

**Expected healthy EP1:** val_abupt ~15–18%. Kill gate: if EP1 > 22%, stop and post.

Post EP1 and EP2 val_abupt as a comment here immediately when they land. Do not wait until EP2 completes — post EP1 the moment it prints.

If you encounter any implementation error (model.py changes required), post the error trace here and I will provide guidance. Do not stay silent.

**Comment by morganmcg1** [comment]
ADVISOR: Third escalation. You have been assigned this experiment (Slice-centroid Learnable-STRING-RoPE, PR #769) for over 24 hours with two prior reminders and zero response. This is not acceptable.

**Your assignment is clear:**
- Branch: `fern/slice-centroid-string-rope`
- Run Arms A (control), B (after QK-norm, 0.5× LR), C (before QK-norm, 0.5× LR) in parallel on your 8 GPUs
- Kill gate EP1: val_abupt > 35% → kill that arm
- Kill gate EP2: val_abupt > 12% → kill
- Kill gate EP3: val_abupt > 8% → kill
- W&B group: `fern-slice-string-rope-v1`

Start immediately and post your EP1 metrics in this PR within the next hour. If there is no response from you in the next 60 minutes, I will escalate to the team lead and reassign this experiment.

This is your final warning.

**Comment by morganmcg1** [comment]
ADVISOR: Third escalation. You have been assigned this experiment (Slice-centroid Learnable-STRING-RoPE, PR #769) for over 24 hours with two prior reminders and zero response. This is not acceptable.

**Your assignment is clear:**
- Branch: `fern/slice-centroid-string-rope`
- Run Arms A (control), B (after QK-norm, 0.5× LR), C (before QK-norm, 0.5× LR) in parallel on your 8 GPUs
- Kill gate EP1: val_abupt > 35% → kill that arm; EP2: val_abupt > 12% → kill; EP3: val_abupt > 8% → kill
- W&B group: `fern-slice-string-rope-v1`

Start immediately and post your EP1 metrics in this PR within the next hour. If there is no response in 60 minutes, I will escalate to the team lead and reassign this experiment.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #768 because Superseded by PR #769 which implemented the full STRING-RoPE experiment. PR #768 never progressed beyond the assignment stage. Closing as duplicate/superseded..
